source/slang/slang-check-decl.cpp

```diff
@@ -3232,6 +3232,16 @@ namespace Slang
 
             _validateCrossModuleInheritance(decl, inheritanceDecl);
         }
+
+        if (decl->findModifier<ComInterfaceAttribute>())
+        {
+            // `associatedtype` declaration is not allowed in a COM interface declaration.
+            for (auto associatedType : decl->getMembersOfType<AssocTypeDecl>())
+            {
+                getSink()->diagnose(
+                    associatedType, Diagnostics::associatedTypeNotAllowInComInterface);
+            }
```

Comment:
I'm going to push a design proposal for the `[COM] interface` feature shortly, but it is worth noting that we could potentially lift this restriction, and doing so could be beneficial in some cases. For now, there is no reason not to have the restriction, though.

It is also worth noting that there are several other kinds of member declarations we probably want to exclude. For example, neither `property` nor `__init` declarations will make a ton of sense inside a `[COM]` interface until we take the time to define rules for them.